File: 0801-0900/0835-Image Overlap/0835-Image Overlap.py
import collections
class Solution:
    def largestOverlap(self, A: List[List[int]], B: List[List[int]]) -> int:
        A1 = [i * len(A) * 2 + j for i in range(len(A)) for j in range(len(A)) if A[i][j]]
        B1 = [i * len(A) * 2 + j for i in range(len(A)) for j in range(len(A)) if B[i][j]]
        table = collections.Counter(i - j for i in A1 for j in B1)
        return max(table.values() or [0])

File: 0801-0900/0835-Image Overlap/0835-Image Overlap.cpp
class Solution {
public:
    int largestOverlap(vector<vector<int>>& A, vector<vector<int>>& B) {
        int n = A.size();
        vector<int> A1, B1;
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                if (A[i][j] == 1) {
                    A1.push_back(i * n * 2 + j);
                }
                
                if (B[i][j] == 1) {
                    B1.push_back(i * n * 2 + j);
                }
            }
        }
        
        unordered_map<int, int> table;
        for (int a : A1) {
            for (int b : B1) {
                ++table[a - b];
            }
        }
        
        int result = 0;
        for (auto t : table) {
            result = max(result, t.second);
        }
            
        return result;
    }
};
